FlxAction API

larsiusprime opened the pull request:
Squashed commits:

[933313d] travis fix for flxactions (+2 squashed commit)

Squashed commit:

[06b7a0e] Added two new action inputs

[b538d6b] style fix (+1 squashed commits)

Squashed commits:

[9ffdfef] small tweak (+1 squashed commits)

Squashed commits:

[5d03851] fix sticky analog action x/y (+1 squashed commits)

Squashed commits:

[8f2d9f1] relative mouse analog input (+1 squashed commits)

Squashed commits:

[2e379e2] fix analog inputs (+1 squashed commits)

Squashed commits:

[70e4214] Add generic IFlxInput digital action input, make FlxActionInput implement IFlxDestroyable (+1 squashed commits)

Squashed commits:

[264a6a9] Fix some bugs with the steam controller (+1 squashed commits)

Squashed commits:

[463dc05] avoid double updates on check (+1 squashed commits)

Squashed commits:

[9b9c2f0] Minor docs fix. (Analog actions are always analog!) (+1 squashed commits)

Squashed commits:

[8220d05] put in FLX_NO_MOUSE, FLX_NO_GAMEPAD, FLX_NO_KEYBOARD checks... (+1 squashed commits)

Squashed commits:

[e6e6d7f] make compiler happy if !steamwrap (+1 squashed commits)

Squashed commits:

[2ec051e] Compile fix for travis (+1 squashed commits)

Squashed commits:

[d44d3c0] finish up action stuff (+4 squashed commit)

Squashed commit:

[fb923a4] Getting close to final version of FlxAction stuff

[58fcfba] more cleanup

[750f177] cleanup

[f15193f] Flixel Actions -- first commit (+2 squashed commit)

Squashed commit:

[9025228] cleanup

[a516cfd] Flixel Actions -- first commit

MSGhero commented:
Was this the result of the github squash PR?

larsiusprime commented:
Nope, the regular kind

larsiusprime commented:
@Gama11 : The code climate reports seem to have a lot of false positives related to imports; if I remove some of the imports it mentions (particularly the "same package" ones) then the code won't compile any more and fail both locally and in Travis.

The rest seem fine; the "unused import detected" seems more reliable but totally trips up over conditionals (like #if steamwrap)

larsiusprime commented:
@Gama11: Travis is failing (only on neko), but it doesn't look to be my fault:
```
haxelib run openfl test unit neko

bin/linux64/neko/haxe/ApplicationMain.hx:202: characters 57-75 : Warning : Use programPath instead

bin/linux64/neko/haxe/ApplicationMain.hx:86: characters 42-60 : Warning : Use programPath instead

Called from /home/travis/haxe/std/neko/_std/sys/ssl/Certificate.hx line 120

Uncaught exception - load.c(237) : Failed to load library : ssl.ndll (ssl.ndll: cannot open shared object file: No such file or directory)
```

And the only things code climate is catching is imports. Let me know if there's anything more to look into here before final review.

Beeblerox commented on flixel/input/actions/FlxAction.hx:
```diff
@@ -0,0 +1,284 @@
+	
+	private function addGenericInput(input:FlxActionInput):FlxAction
+	{
+		if (false == checkExists(input))
```
this could be just `if (!checkExists(input))`

Beeblerox commented on flixel/input/actions/FlxAction.hx:
```diff
@@ -0,0 +1,284 @@
+		_check = false;
+		
+		var len = inputs.length;
+		for (i in 0...len)
```
there is another trick for backward iteration through array:
```
for (i in -(len - 1)...0)
{
    var j = -i;
    var input = inputs[j];
    ...
```
Just for the info. Pick up the way you like more

Beeblerox replied:
or you can use while loop

larsiusprime replied:
done

Beeblerox commented on flixel/input/actions/FlxActionInputAnalog.hx:
```diff
@@ -0,0 +1,367 @@
+		super(FlxInputDevice.MOUSE, -1, cast Trigger, Axis);
+	}
+	
+	override public function destroy():Void 
```
if you just call `super.destroy()` here, then there is no need to override `destroy()`

larsiusprime replied:
Must have been left over from debug breakpoint, removing

Beeblerox commented on flixel/input/actions/FlxActionInputDigital.hx:
```diff
@@ -0,0 +1,294 @@
+	}
+}
+
+class FlxActionInputDigitalIFlxInput extends FlxActionInputDigital
```
don't know how to feel about this class name (there are 2 `Flx` in it).

MSGhero replied:
That's kinda hilarious

Gama11 replied:
...and two "input". :D

larsiusprime replied:
Heh. What's funny is that this is a direct result of the naming convention I've established and it just turns out to be a really ridiculous edge case.

"a digital input for a flxAction, where the input is of type IFlxInput"

I'm not entirely sure whether I should rework the entire naming convention because "FlxActionInputDigital" gets a bit unwieldy, or what. 

Is it our policy to stick "Flx" on literally everything, or drop it for certain situations? And are we at all consistent in that? It'd be painful to remove it for existing classes, but what about new ones? `ActionInputDigital<Something>` would certainly be a bit less unwieldy.

However, it doesn't look easy to drop "Action" because without it "Input" becomes ambiguous because there's so many other contexts in which "Input" is used.

Lastly, I do think it's important to distinguish between Digital and Analog action inputs, but maybe that should come at the beginning rather than squashed towards the end?

`DigitalActionInput<Something>`

Anyways, open to ideas.

Gama11 replied:
I don't like the `Flx`, but we are at least sort of conistent with it right now.. Changing that would make things worse I feel like.

gamedevsam replied:
Why isn't all this logic just contained inside ```FlxActionInputDigital```? That way we avoid the need for the insanely long class name :-P

Beeblerox commented on flixel/input/actions/FlxActionManager.hx:
```diff
@@ -0,0 +1,735 @@
+	 */
+	public function addSet(set:FlxActionSet):Int
+	{
+		for (s in sets)
```
this check could be reduced to:
```
if (sets.indexOf(set) >= 0)
    return -1;
```

Gama11 replied:
Should actually use `contains()` instead of `indexOf()` from `FlxArrayUtil` via static extension.

Beeblerox replied:
sure

larsiusprime replied:
on it

Beeblerox commented on flixel/input/actions/FlxActionManager.hx:
```diff
@@ -0,0 +1,735 @@
+	
+	public function destroy():Void
+	{
+		FlxDestroyUtil.destroyArray(sets);
```
this line should be `sets = FlxDestroyUtil.destroyArray(sets);`

larsiusprime replied:
done.

Beeblerox commented on flixel/input/actions/FlxActionManager.hx:
```diff
@@ -0,0 +1,735 @@
+		FlxDestroyUtil.destroyArray(sets);
+		register.destroy();
+		register = null;
+		sets = null;
```
and then this line should be deleted

larsiusprime replied:
I think you mean these two? But just to check:

```haxe
register = null;
sets = null;
```

Beeblerox commented on flixel/input/actions/FlxActionManager.hx:
```diff
@@ -0,0 +1,735 @@
+	public function destroy():Void
+	{
+		FlxDestroyUtil.destroyArray(sets);
+		register.destroy();
```
same for this line:
```
register = FlxDestroyUtil.destroy(register);
```

larsiusprime replied:
done.

Beeblerox commented on flixel/input/actions/FlxActionManager.hx:
```diff
@@ -0,0 +1,735 @@
+		{
+			if (addSet(FlxActionSet.fromJSON(set, CallbackDigital, CallbackAnalog)) != -1)
+			{
+				trace("added set : " + sets[sets.length - 1].name);
```
i believe that this line should be deleted

larsiusprime replied:
done.

Beeblerox commented on flixel/input/actions/FlxActionManager.hx:
```diff
@@ -0,0 +1,735 @@
+	public function removeSet(Set:FlxActionSet, Destroy:Bool = true):Bool
+	{
+		var success = false;
+		for (i in 0...sets.length)
```
this also could be changed by using indexOf:
```
var index:Int = sets.indexOf(Set);
if (index >= 0)
{
    success = true;
    sets.splice(index, 1);
    if (Destroy)
        FlxDestroyUtil.destroy(Set);
}
```

Gama11 replied:
Should actually use `contains()` instead of `indexOf()` from `FlxArrayUtil` via static extension.

Beeblerox replied:
or you can use array's remove() method:
```
var result:Bool = sets.remove(Set);
if (result)
{
    if (Destroy)
        FlxDestroyUtil.destroy(Set);

    onChange();
}
return result;
```

larsiusprime replied:
used Beeblerox's message

Beeblerox commented on flixel/input/actions/FlxActionSet.hx:
```diff
@@ -0,0 +1,322 @@
+	 */
+	public function addDigital(Action:FlxActionDigital):Bool
+	{
+		if (digitalActions.indexOf(Action) != -1) return false;
```
as @Gama11 mentioned you can change this line to:
```
if (FlxArrayUtil.contains(digitalActions, Action)) return false;
```

Gama11 replied:
(Via static extension though)
Am 14.05.2016 13:36 schrieb "Zaphod" <notifications@github.com>:

> In flixel/input/actions/FlxActionSet.hx
> <https://github.com/HaxeFlixel/flixel/pull/1805#discussion_r63274929>:
>
> > +	 * @param	Attach	true: adds inputs, false: removes inputs
> > +	 */
> > +	public function attachSteamController(Handle:Int, Attach:Bool = true):Void
> > +	{
> > +		attachSteamControllerSub(Handle, Attach, FlxInputType.DIGITAL, digitalActions, null);
> > +		attachSteamControllerSub(Handle, Attach, FlxInputType.ANALOG, null, analogActions);
> > +	}
> > +	
> > +	/**
> > +	 * Add a digital action to this set if it doesn't already exist
> > +	 * @param	Action a FlxActionDigital
> > +	 * @return	whether it was added
> > +	 */
> > +	public function addDigital(Action:FlxActionDigital):Bool
> > +	{
> > +		if (digitalActions.indexOf(Action) != -1) return false;
>
> as @Gama11 <https://github.com/Gama11> mentioned you can change this line
> to:
>
> if (FlxArrayUtil.contains(digitalActions, Action)) return false;
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/HaxeFlixel/flixel/pull/1805/files/d62bf2c0c11d1dec8da3c05d5c29fbac247ef1ad#r63274929>
>

larsiusprime replied:
done

Beeblerox commented on flixel/input/actions/FlxActionSet.hx:
```diff
@@ -0,0 +1,322 @@
+	 */
+	public function addAnalog(Action:FlxActionAnalog):Bool
+	{
+		if (analogActions.indexOf(Action) != -1) return false;
```
same here

larsiusprime replied:
done

Beeblerox commented on flixel/input/actions/FlxActionSet.hx:
```diff
@@ -0,0 +1,322 @@
+	}
+	
+	public function destroy():Void
+	{
```
this method could be reduced to 2 lines:
```
digitalActions = FlxDestroyUtil.destroyArray(digitalActions);
analogActions = FlxDestroyUtil.destroyArray(analogActions);
```

larsiusprime replied:
done

Beeblerox commented on flixel/input/actions/FlxActionSet.hx:
```diff
@@ -0,0 +1,322 @@
+	 */
+	public function removeDigital(Action:FlxActionDigital, Destroy:Bool = true):Bool
+	{
+		var index = digitalActions.indexOf(Action);
```
you can use `remove()` here:
```
var result = digitalActions.remove(Action);
if (result && Destroy)
{
    Action.destroy();
}
return result;
```

larsiusprime replied:
done

Beeblerox commented on flixel/input/actions/FlxActionSet.hx:
```diff
@@ -0,0 +1,322 @@
+	 */
+	public function removeAnalog(Action:FlxActionAnalog, Destroy:Bool = true):Bool
+	{
+		var index = analogActions.indexOf(Action);
```
same here

larsiusprime replied:
done

Beeblerox commented on flixel/input/actions/FlxActionSet.hx:
```diff
@@ -0,0 +1,322 @@
+						}
+					}
+				}
+				else
```
you can change following lines to:
```
else if (inputExists)
{
    action.removeInput(theInput);
}
```
to avoid nested conditional (unless you plan to add some other logic there)

larsiusprime replied:
done

Beeblerox commented on flixel/input/actions/FlxAction.hx:
```diff
@@ -0,0 +1,284 @@
+	
+	private function checkExists(input:FlxActionInput):Bool
+	{
+		for (i in inputs)
```
you can change this to just one line:
```
return FlxArrayUtil.contains(inputs, input);
```

Gama11 replied:
Should be `inputs.contains(input)`

larsiusprime replied:
done

Beeblerox commented on flixel/input/actions/FlxSteamController.hx:
```diff
@@ -0,0 +1,242 @@
+		return data;
+	}
+	#else
+	private static function getAnalogActionData(controller:Int, action:Int):Dynamic
```
maybe add here third optional argument `?data:Dynamic` for better match with steamwrap version of the method?

larsiusprime replied:
sure

larsiusprime commented:
Most of the codeclimate checks seem like false positives, but I'm checking in on what's up with travis.

larsiusprime commented:
Keep in mind I have a demo for FlxActions, too:
https://github.com/larsiusprime/flixel-demos/tree/actions_/Input/FlxAction

Let me know if we're approaching mergability and I'll look into our squash options...

larsiusprime commented:
On the checks:
- codeclimate seems to be catching a lot of false positives here. I addressed all that made sense to me.
- travis might have false positives too? Some cpp/neko tests are failing but they seem green internally, I can't find any actionable error messages that I understand.

MSGhero commented on flixel/input/actions/FlxActionInputAnalog.hx:
```diff
@@ -0,0 +1,362 @@
+import flixel.input.mouse.FlxMouseButton;
+
+#if steamwrap
+import steamwrap.api.Steam;
```
I don't see what these steam imports are for?

larsiusprime replied:
https://github.com/larsiusprime/haxeflixel/blob/f40ce711a3e9ad6c3e637b682ca840a9b76f13d0/flixel/input/actions/FlxActionInputAnalog.hx#L229

Steam controller!

MSGhero replied:
Right, but nowhere in this file specifically are you using those classes. FlxSteamController should have those imports, but not here, and that's what haxe-syntax is catching.

MSGhero commented:
Your travis is fine, every PR is having those issues.

Tembac commented:
Is there any estimation on when is this going to be implemented on HaxeFlixel?

@larsiusprime article about this is great: http://www.gamasutra.com/blogs/LarsDoucet/20161012/283057/Steam_Dev_Days_Steam_Controller.php

larsiusprime commented:
@Gama11 Trying to finish this up at long last. Travis is all green after re-running the tests, but codeclimate's analysis has expired and I don't remember how to re-run it. IIRC, we had fixed everything except for the really nitpicky stuff.

larsiusprime commented:
Just ran checkstyle locally on this branch, results (removing everything from the rest of flixel irrelevant to this PR), no errors, all warnings:

```
../../flixel/input/actions/FlxAction.hx:2: character 0 : Warning: Empty line required after package declaration
../../flixel/input/actions/FlxAction.hx:23: character 0 : Warning: Wrong indentation character (should be tab)
../../flixel/input/actions/FlxAction.hx:83: character 0 : Warning: Wrong indentation character (should be tab)
../../flixel/input/actions/FlxAction.hx:2: characters 0-6 : Warning: Unused import "flixel.util.FlxArrayUtil" detected
../../flixel/input/actions/FlxAction.hx:3: characters 0-6 : Warning: Detected import "flixel.input.actions.FlxActionInput" from same package "flixel.input.actions"
../../flixel/input/actions/FlxAction.hx:4: characters 0-6 : Warning: Detected import "flixel.input.actions.FlxActionInputAnalog" from same package "flixel.input.actions"
../../flixel/input/actions/FlxAction.hx:5: characters 0-6 : Warning: Detected import "flixel.input.actions.FlxActionInputDigital" from same package "flixel.input.actions"
../../flixel/input/actions/FlxActionInput.hx:3: characters 0-6 : Warning: Unused import "flixel.input.FlxInput" detected
../../flixel/input/actions/FlxActionInput.hx:4: characters 0-6 : Warning: Detected import "flixel.input.actions.FlxAction" from same package "flixel.input.actions"
../../flixel/input/actions/FlxActionInputAnalog.hx:298: character 0 : Warning: Wrong indentation character (should be tab)
../../flixel/input/actions/FlxActionInputAnalog.hx:299: character 0 : Warning: Wrong indentation character (should be tab)
../../flixel/input/actions/FlxActionInputAnalog.hx:307: character 0 : Warning: Wrong indentation character (should be tab)
../../flixel/input/actions/FlxActionInputAnalog.hx:308: character 0 : Warning: Wrong indentation character (should be tab)
../../flixel/input/actions/FlxActionInputAnalog.hx:309: character 0 : Warning: Wrong indentation character (should be tab)
../../flixel/input/actions/FlxActionInputAnalog.hx:311: character 0 : Warning: Wrong indentation character (should be tab)
../../flixel/input/actions/FlxActionInputAnalog.hx:4: characters 0-6 : Warning: Detected import "flixel.input.actions.FlxAction" from same package "flixel.input.actions"
../../flixel/input/actions/FlxActionInputAnalog.hx:5: characters 0-6 : Warning: Detected import "flixel.input.actions.FlxActionInput" from same package "flixel.input.actions"
../../flixel/input/actions/FlxActionInputAnalog.hx:8: characters 0-6 : Warning: Unused import "flixel.input.mouse.FlxMouseButton" detected
../../flixel/input/actions/FlxActionInputAnalog.hx:11: characters 0-6 : Warning: Unused import "steamwrap.api.Steam" detected
../../flixel/input/actions/FlxActionInputAnalog.hx:12: characters 0-6 : Warning: Unused import "steamwrap.api.Controller" detected
../../flixel/input/actions/FlxActionInputDigital.hx:292: character 0 : Warning: Too many consecutive empty lines (> 1)
../../flixel/input/actions/FlxActionInputDigital.hx:5: characters 0-6 : Warning: Detected import "flixel.input.actions.FlxAction" from same package "flixel.input.actions"
../../flixel/input/actions/FlxActionInputDigital.hx:6: characters 0-6 : Warning: Detected import "flixel.input.actions.FlxActionInput" from same package "flixel.input.actions"
../../flixel/input/actions/FlxActionInputDigital.hx:13: characters 0-6 : Warning: Unused import "steamwrap.api.Steam" detected
../../flixel/input/actions/FlxActionInputDigital.hx:14: characters 0-6 : Warning: Unused import "steamwrap.api.Controller" detected
../../flixel/input/actions/FlxActionManager.hx:22: character 0 : Warning: Empty line not allowed after comment(s)
../../flixel/input/actions/FlxActionManager.hx:22: character 0 : Warning: Too many consecutive empty lines (> 1)
../../flixel/input/actions/FlxActionManager.hx:255: character 0 : Warning: Too many consecutive empty lines (> 1)
../../flixel/input/actions/FlxActionManager.hx:572: character 0 : Warning: Empty line not allowed after comment(s)
../../flixel/input/actions/FlxActionManager.hx:719: character 0 : Warning: Too many consecutive empty lines (> 1)
../../flixel/input/actions/FlxActionManager.hx:145: character 0 : Warning: Wrong indentation character (should be tab)
../../flixel/input/actions/FlxActionManager.hx:489: character 0 : Warning: Wrong indentation character (should be tab)
../../flixel/input/actions/FlxActionManager.hx:503: character 0 : Warning: Wrong indentation character (should be tab)
../../flixel/input/actions/FlxActionManager.hx:270: character 8 : Warning: No space between "switch" and "("
../../flixel/input/actions/FlxActionManager.hx:275: character 10 : Warning: No space between "switch" and "("
../../flixel/input/actions/FlxActionManager.hx:290: character 10 : Warning: No space between "switch" and "("
../../flixel/input/actions/FlxActionManager.hx:301: character 10 : Warning: No space between "switch" and "("
../../flixel/input/actions/FlxActionManager.hx:373: characters 89-90 : Warning: Left curly should be on new line (only whitespace before curly)
../../flixel/input/actions/FlxActionManager.hx:392: character 3 : Warning: Right curly should be alone on a new line
../../flixel/input/actions/FlxActionManager.hx:9: characters 0-6 : Warning: Unused import "flixel.input.FlxInput" detected
../../flixel/input/actions/FlxActionManager.hx:10: characters 0-6 : Warning: Unused import "flixel.input.IFlxInput" detected
../../flixel/input/actions/FlxActionManager.hx:11: characters 0-6 : Warning: Detected import "flixel.input.actions.FlxAction" from same package "flixel.input.actions"
../../flixel/input/actions/FlxActionManager.hx:12: characters 0-6 : Warning: Unused import "flixel.input.mouse.FlxMouseButton" detected
../../flixel/input/actions/FlxActionManager.hx:13: characters 0-6 : Warning: Unused import "flixel.input.gamepad.FlxGamepadInputID" detected
../../flixel/input/actions/FlxActionManager.hx:17: characters 0-6 : Warning: Unused import "steamwrap.api.Controller" detected
../../flixel/input/actions/FlxActionManager.hx:18: characters 0-6 : Warning: Unused import "steamwrap.api.Steam" detected
../../flixel/input/actions/FlxActionSet.hx:302: character 11 : Warning: No space between "if" and "("
../../flixel/input/actions/FlxActionSet.hx:3: characters 0-6 : Warning: Unused import "flixel.input.FlxInput" detected
../../flixel/input/actions/FlxActionSet.hx:6: characters 0-6 : Warning: Detected import "flixel.input.actions.FlxAction" from same package "flixel.input.actions"
../../flixel/input/actions/FlxActionSet.hx:7: characters 0-6 : Warning: Detected import "flixel.input.actions.FlxActionInput" from same package "flixel.input.actions"
../../flixel/input/actions/FlxActionSet.hx:8: characters 0-6 : Warning: Detected import "flixel.input.actions.FlxActionInputAnalog" from same package "flixel.input.actions"
../../flixel/input/actions/FlxActionSet.hx:9: characters 0-6 : Warning: Detected import "flixel.input.actions.FlxActionInputDigital" from same package "flixel.input.actions"
../../flixel/input/actions/FlxActionSet.hx:13: characters 0-6 : Warning: Unused import "steamwrap.api.Steam" detected
../../flixel/input/actions/FlxSteamController.hx:2: character 0 : Warning: Empty line required after package declaration
../../flixel/input/actions/FlxSteamController.hx:104: character 0 : Warning: Too many consecutive empty lines (> 1)
../../flixel/input/actions/FlxSteamController.hx:156: character 0 : Warning: Too many consecutive empty lines (> 1)
../../flixel/input/actions/FlxSteamController.hx:3: characters 0-6 : Warning: Detected import "flixel.input.actions.FlxActionInput" from same package "flixel.input.actions"
../../flixel/input/actions/FlxSteamController.hx:6: characters 0-6 : Warning: Unused import "steamwrap.api.Controller" detected
```

larsiusprime commented:
Latest fix addresses all of the issues mentioned by code climate pertaining to FlxActions.

All that remains when running it locally on this branch deal with other classes which have presumably already been fixed upstream, or at least are not my fault/problem:

```
Total Issues: 20 (Errors: 0, Warnings: 20, Infos: 0)
../../flixel/FlxGame.hx:3: characters 0-6 : Warning: Unused import "flash.display.Bitmap" detected
../../flixel/FlxGame.hx:4: characters 0-6 : Warning: Unused import "flash.display.BitmapData" detected
../../flixel/FlxGame.hx:10: characters 0-6 : Warning: Unused import "flash.geom.ColorTransform" detected
../../flixel/FlxGame.hx:11: characters 0-6 : Warning: Unused import "flash.geom.Matrix" detected
../../flixel/FlxGame.hx:15: characters 0-6 : Warning: Unused import "flixel.math.FlxAngle" detected
../../flixel/FlxGame.hx:20: characters 0-6 : Warning: Unused import "flixel.util.FlxColor" detected
../../flixel/FlxGame.hx:21: characters 0-6 : Warning: Unused import "flixel.util.FlxDestroyUtil" detected
../../flixel/input/mouse/FlxMouse.hx:17: characters 0-6 : Warning: Detected import "flixel.input.mouse.FlxMouseButton" from same package "flixel.input.mouse"
../../flixel/math/FlxAngle.hx:9: characters 0-6 : Warning: Detected import "flixel.math.FlxPoint" from same package "flixel.math"
../../flixel/math/FlxAngle.hx:10: characters 0-6 : Warning: Unused import "flixel.system.macros.FlxMacroUtil" detected
../../flixel/system/FlxSound.hx:19: characters 0-6 : Warning: Unused import "flixel.system.frontEnds.SoundFrontEnd" detected
../../flixel/system/debug/console/Console.hx:4: characters 0-6 : Warning: Unused import "openfl.events.Event" detected
../../flixel/system/debug/console/Console.hx:14: characters 0-6 : Warning: Unused import "flixel.system.debug.FlxDebugger" detected
../../flixel/system/debug/watch/Tracker.hx:14: characters 0-6 : Warning: Unused import "flixel.FlxSubState" detected
../../flixel/system/debug/watch/Tracker.hx:16: characters 0-6 : Warning: Unused import "flixel.group.FlxGroup" detected
../../flixel/system/debug/watch/Tracker.hx:25: characters 0-6 : Warning: Unused import "flixel.ui.FlxButton" detected
../../flixel/system/debug/watch/Tracker.hx:31: characters 0-6 : Warning: Unused import "flixel.effects.particles.FlxEmitter" detected
../unit/src/flixel/system/FlxPreloaderTest.hx:15: character 0 : Warning: Wrong indentation character (should be tab)
../unit/src/flixel/system/FlxPreloaderTest.hx:16: character 0 : Warning: Wrong indentation character (should be tab)
../unit/src/flixel/system/FlxPreloaderTest.hx:17: character 0 : Warning: Wrong indentation character (should be tab)
Build succeeded
Done(0)
```

Will do my best to fix any new issues popping up until both Travis and Codeclimate are green as far as my changes are concerned. Then it's MERGE TIME.

larsiusprime commented:
Okay, finally got all the checks to pass and fixed all the merge errors. Ready for final review, @Gama11!

Additionally tested against Defender's Quest and the FlxActions demo (which will be merged into flixel-demos after this goes through).

larsiusprime commented:
Anything else required here @Gama11 and @Beeblerox ?

Gama11 commented on flixel/input/actions/FlxActionInputAnalog.hx:
```diff
@@ -0,0 +1,373 @@
+	
+	override public function update():Void 
+	{
+		if (deviceID != FlxInputDeviceID.ALL)	//analog data is only meaningful on an individual device
```
Could reduce nesting a bit by inverting this if.

larsiusprime replied:
Okay, I replaced it with:

```haxe
if (deviceID == FlxInputDeviceID.ALL)
{
	return; //analog data is only meaningful on an individual device
}
```

And removed the later block. Is that what you meant?

Gama11 commented on flixel/input/actions/FlxAction.hx:
```diff
@@ -0,0 +1,353 @@
+		_x = null;
+		_y = null;
+		
+		if (_timestamp == @:privateAccess FlxG.game._total)
```
Why is `@:privateAccess` to `FlxG.game._total` necessary here? Not a fan of that. Isn't `ticks` sufficient?

larsiusprime replied:
Somehow I was unaware of ticks. Fixing.

IBwWG commented:
Are there tests?

IBwWG commented:
I'd like to try out the FlxAction demo on Linux64.  So, I installed the flixel/actions__ branch and the flixel-demos/actions_ branch, and tried to compile the FlxAction demo.  Firstly, the versions of OpenFL and Lime I had would not let me compile it because of the deprecated Tilesheet in OpenFL 4.

openfl 3.6.1 + lime 3.7.2 finally let me compile, but launching crashes immediately with `Called from openfl._legacy.utils.ByteArray::__init__ openfl/_legacy/utils/ByteArray.hx line 761`

openfl 3.6.1 + lime 2.9.1 compiles too, but ibid.

How are y'all compiling this exactly?  :)

Gama11 commented:
>openfl 3.6.1 + lime 3.7.2

That's a weird combination, are those even compatible?

IBwWG commented:
@Gama11 That one was mostly by accident, but I included it because it compiled anyway and didn't complain (unlike incompatible openfl 4.x and lime 3.x versions, which do.)

IBwWG commented:
I mean, openfl itself didn't complain.  The demo still crashed.

IBwWG commented:
@Gama11 which OS do you primarily compile on?

IBwWG commented:
Interesting, I just tried it on Windows with flash target (openfl 3.6.1, lime 2.9.1) and the demo works fine.

On Linux I am targeting cpp though.  (I don't have a working flash standalone player atm.)

IBwWG commented:
Interesting, Windows cpp/debug worked too.  Both systems have the same hxcpp version, 3.4.43.

IBwWG commented:
OK, this is strange, I got it to compile.  I had been running the wrong binary.  But where did *that* one come from?

`export/linux64/cpp/` contains this strange arrangement:

`bin  debug  haxe  obj  release`

...and then `debug` and `release` each contain the usual `bin haxe obj`.  I had been running `debug/bin/`'s copy.  Why did it decide to switch output directories?

First time I got it to compile (with the strange combination you pointed out @Gama11 although it doesn't produce a warning):

```
...FlxAction/export/linux64/cpp/release/bin $ ./FlxAction 
Error : Null Function Pointer
```

The second time I got it to compile (probably with lime 2.9.1) it's now hard to tell because I remembered there are the debug/release directories, and that I probably need to launch from the game's own directory for it to work, so I opened a second terminal, and then it gets harder to trace what happened.  Also figuring I could get a stack trace, I modified the demo quickly to use crashdumper to help me debug, and started compiling with `openfl -debug build linux`.

But there's still something wrong here, I went to lime 2.9.1 and built for release, without crashdumper, and got the same error as on the original compile again.

IBwWG commented:
Oh my, what is happening here.  Now on debug, sans crashdumper, with lime 2.9.1, which is what worked a moment ago, I `rm -rf` the `cpp` directory and do a fresh compile, and get null function pointer.

IBwWG commented:
OK in this case it's because I tried to run it not from the directory (i.e. from the grandparent, ran `cpp/bin/FlxAction`.  So that *was* important to be doing before.

IBwWG commented:
I think the different output spots (and hence my confusion as to whether to be under `cpp/debug/bin` or `cpp/bin` when launching) are due to how this is handled differently between lime versions.

Wow, a lot of little changes can really add up to some confusion sometimes.

All that to say...cool demo :)

Eiyeron commented:
(Well, for the thread spam, there is an edit button, it might help reducing the post cluttering.)
I look forward to this change, maybe sometimes it'll be merged. I wonder if a tool such as ReWired could eventually raise from such PR. I haven't looked deep in the sources but I wonder if managing controllers on the fly (plugging in/out) or reassignation might be supported or become the subject of a wrapper.

IBwWG commented:
Indeed @Eiyeron thanks for the suggestion.  I've cleaned everything up.  :)

Gama11 commented on flixel/input/actions/FlxActionManager.hx:
```diff
@@ -279,7 +279,7 @@ class FlxActionManager implements IFlxInputManager implements IFlxDestroyable
 		
 		var i:Int = 0;
-		var actionSets:Array<Dynamic> = Reflect.hasField(data, "actionSets") ? Reflect.field(data, "actionSets") : null;
+		var actionSets:Array<ActionSetJSON> = Reflect.hasField(data, "actionSets") ? Reflect.field(data, "actionSets") : null;
```
The reflection shouldn't be necessary anymore with the typedef?

larsiusprime commented:
I believe I have now fixed all but 4 code climate issues, which all look like nitpicky false positives (if I "fix" them the code won't compile).

On my machine, the unit tests pass for both `legacy` and `next`. The only test that failed was one that I haven't touched (something in `FlxButtonTest`, `testTriggerAnimationOnce`), so I don't see that as relevant to my PR.

larsiusprime commented:
Okay, technically this cleanup remains:

1) Steamwrap is incompatible with neko, need proper guards there
2) Steamwrap appears incompatible with haxe 3.2, need proper guards there
3) Not all of my gamepad spoofing for test purposes works in flash, need proper guards there

larsiusprime commented:
Hmm, seems like C++ with haxe development fails too? That's really weird, doesn't happen locally.

Gama11 commented:
This doesn't look good, could be a Haxe / cpp bug: https://travis-ci.org/HaxeFlixel/flixel/jobs/217374319#L789-L790

Could also because we're using outdated Haxe nightlies with latest hxcpp, though (see https://github.com/HaxeFoundation/haxe/issues/6149).

larsiusprime commented:
@Gama11 I'm all for that, I just tried it before and it didn't seem to work. I figure I'll just get the tests to light up green first, then find/replace to clean up the verbosity once I understand FlxDefines better

larsiusprime commented:
I've got it passing on everything but cpp + haxe dev version. Should have this all green soon!

gamedevsam commented on flixel/input/actions/FlxAction.hx:
```diff
@@ -0,0 +1,352 @@
+	private var _y:Null<Float> = null;
+	
+	private var _timestamp:Int = 0;
+	private var _check:Bool = false;
```
I think a better name for this variable is ```_checked```

gamedevsam commented on flixel/input/actions/FlxSteamController.hx:
```diff
@@ -0,0 +1,342 @@
+	}
+}
+
+typedef DigitalActionData =
```
Would prefer if this was a class. Removes any doubt whether it will result in fast native code.

gamedevsam commented on flixel/input/gamepad/FlxGamepadManager.hx:
```diff
@@ -390,12 +411,14 @@ class FlxGamepadManager implements IFlxInputManager
 		if (Device == null)
 			return;
-
+		
```
Unnecessary whitespace

gamedevsam commented on flixel/input/gamepad/FlxGamepadManager.hx:
```diff
@@ -303,6 +319,8 @@ class FlxGamepadManager implements IFlxInputManager
 	@:allow(flixel.FlxG)
 	private function new() 
 	{
+		deviceConnected = new FlxTypedSignal<FlxGamepad->Void>();
```
I wonder if these abstract ```@:multiType``` signals actually generate fast C++ code. Usually generic code must be marked with ```@:generic``` and you have to use classes that perform the signal action in order to get the benefits from generics.

MSGhero commented:
Oh snap, is it merge time?

larsiusprime commented:
Not quite, got some thorny travis issues to chase down. Any help would be appreciated.

larsiusprime commented:
Near as I can tell we have exactly one failing test:
```
Class: flixel.input.actions.FlxActionSetTest ..x....
    ERROR: massive.munit.UnhandledException: Could not load module steamwrap@SteamWrap_ShowAnalogActionOrigins__prime at flixel.input.actions.FlxActionSetTest#testAttachSteamController
```

I just updated steamwrap to support 64-bits by default just to be safe, and I'm running the unit tests locally to see what comes up and if I get the same or different results as travis.

larsiusprime commented:
Good news everyone!

I get the exact same behavior when running the tests locally.

```
Class: flixel.input.actions.FlxActionSetTest ..x....
    ERROR: massive.munit.UnhandledException: Could not load module steamwrap@SteamWrap_ShowAnalogActionOrigins__prime at flixel.input.actions.FlxActionSetTest#testAttachSteamController
```
At least this means it's not due to any weird state on travis, and I don't have to wait for it to update to get a diagnosis on whether I've fixed it.

larsiusprime commented:
Commenting out that one test and re-running locally gives me 388/388 passed tests, so I can confirm there are no hidden cascading failures. This looks to be the only remaining issue.

larsiusprime commented:
Found the cause.

When you call this function:
https://github.com/larsiusprime/haxeflixel/blob/actions__/tests/unit/src/flixel/input/actions/FlxActionSetTest.hx#L276

`SteamMock` gets called, which is the true culprit. Anything that calls SteamMock basically triggers this error.
https://github.com/larsiusprime/haxeflixel/blob/actions__/tests/unit/src/flixel/input/actions/FlxActionSetTest.hx#L366

This is the line in SteamMock that causes the error -- initializing a `FakeController`
https://github.com/larsiusprime/haxeflixel/blob/actions__/tests/unit/src/flixel/input/actions/SteamMock.hx#L37

The issue is that `FakeController` extends SteamWrap's `Controller`, which will trigger CFII prime load calls.

https://github.com/larsiusprime/haxeflixel/blob/actions__/tests/unit/src/flixel/input/actions/SteamMock.hx#L123

The one it chokes on is `SteamWrap_ShowAnalogActionOrigins`:
https://github.com/larsiusprime/SteamWrap/blob/master/steamwrap/api/Controller.hx#L534

But as far as I can tell, there's nothing special about it. It seems to just start loading from the last line upwards. If I comment out `SteamWrap_ShowAnalogActionOrigins` it chokes on `SteamWrap_ShowDigitalActionOrigins` and so on up the line.

So I think the true cause of the error is:
Something about this environment causes CFFI Prime loads to fail.

I don't think I actually NEED CFFI Prime functionality since this is a mocked up unit test anyways. Possible workarounds:

- Modify SteamWrap functions to take IController instead of Controller, and make my FakeController implement that interface. This way no CFFI stuff is ever being called.
- Override initialization logic in FakeController to skip the CFFI loading.
- Make CFFI loading work in this environment (probably not worth it).

larsiusprime commented:
Okay, so.... after making sure I had the right version of steamwrap installed, I got it working without having to do anything weird, even from the testing environment!

The trick is that steamwrap has to be built first, and I don't think the testing environment is building it. Probably can work around this by pushing up to date linux binaries maybe...?

larsiusprime commented:
Okay, we're green. Requesting final review, @Gama11 and @Beeblerox

larsiusprime commented:
Updated the flixel-demos PR accordingly:
https://github.com/HaxeFlixel/flixel-demos/pull/268

larsiusprime commented:
@aurel300 -- did I ever send you the bounty for this? Regardless of whether it's merged in today or not I want to make sure you get your bounty.

larsiusprime commented:
There seems to be a crash on neko, but it doesn't look to be this PR's fault, and it seems quite random? The last one passed fine.

Gama11 commented on flixel/input/gamepad/FlxGamepadButton.hx:
```diff
@@ -14,6 +14,8 @@ class FlxGamepadButton extends FlxInput<Int>
 	
 	override public function release():Void
 	{
+		trace("release(" + ID + ")");
```
These traces shouldn't be here.

larsiusprime replied:
How'd those get there?

Gama11 commented on tests/unit/src/flixel/input/actions/SteamMock.hx:
```diff
@@ -0,0 +1,227 @@
+		
+		var key = controller + "_" + actionSet + "_" + action;
+		
+		trace("setDigitalActionOrigins() key = " + key);
```
Is this leftover from debugging?

larsiusprime replied:
Must be. Big long PR's ya know.

larsiusprime commented:
RE: failed tests -- 
1) Codeclimate is complaining about an unused import that is actually required, as far as I can tell
2) Travis is randomly failing on neko for no good reason again. Everything else passes.

Aurel300 commented:
Hurray :)

larsiusprime commented:
@Aurel300 -- remind me -- did we settle your bounty or not? If not, it's time for me to pay up now that the PR is merged.
